model_api/python/model_api/models/classification.py

```diff
@@ -413,11 +425,204 @@ def get_predecessors(lbl, candidates):
                     output_labels.append(new_lbl)
 
         output_predictions = [
-            (self.label_to_idx[lbl], lbl, label_to_prob[lbl]) for lbl in output_labels
+            (self.label_to_idx[lbl], lbl, label_to_prob[lbl])
+            for lbl in sorted(output_labels)
         ]
         return output_predictions
 
 
+class ProbabilisticLabelsResolver(GreedyLabelsResolver):
+    def __init__(self, hierarchical_config, warmup_cache=True) -> None:
+        super().__init__(hierarchical_config)
+        if warmup_cache:
+            self.label_tree.get_labels_in_topological_order()
+
+    def resolve_labels(self, predictions):
+        """Resolves hierarchical labels and exclusivity based on a list of ScoredLabels (labels with probability).
+
+        The following two steps are taken:
+
+        - selects the most likely label from an exclusive (multiclass) group
+        - removes children of "not-most-likely" (non-max) parents in an exclusive group (top-down approach)
+
+        Args:
+            predictions: a list of tuples (label name, score)
+        """
+        label_to_prob = {}
+        for lbl, score in predictions:
+            label_to_prob[lbl] = score
+
+        return self.__resolve_labels_probabilistic(label_to_prob)
+
+    def __resolve_labels_probabilistic(
+        self,
+        label_to_probability,
+    ):
+        """Resolves hierarchical labels and exclusivity based on a probabilistic label output.
+
+        - selects the most likely (max) label from an exclusive group
+        - removes children of non-max parents in an exclusive group
+
+        See `resolve_labels_probabilistic` for parameter descriptions
+
+        Args:
+            label_to_probability : map from label to float.
+
+        Returns:
+            List of labels with probability
+        """
+        # add (potentially) missing ancestors labels for children with probability 0
+        # this is needed so that suppression of children of non-max exclusive labels works when the exclusive
+        # group has only one member
+        label_to_probability = self._add_missing_ancestors(label_to_probability)
+
+        hard_classification = self._resolve_exclusive_labels(label_to_probability)
+
+        # suppress the output of children of parent nodes that are not the most likely label within their group
+        resolved = self._suppress_descendant_output(hard_classification)
+
+        result = []
+        for lbl, probability in sorted(resolved.items()):
+            if probability > 0:  # only return labels with non-zero probability
+                result.append(
+                    (
+                        self.label_to_idx[lbl],
+                        lbl,
+                        # retain the original probability in the output
+                        probability * label_to_probability.get(lbl, 1.0),
```

Comment:
It's just a question. It seems that probability should be either 1.0 or 0.0. If it's True, is it needed to multiply `probability`?

Reply:
You're right, that was done for verbosity in the original SC SDK.